src/agent/src/image_rpc.rs

```diff
@@ -44,6 +70,184 @@ impl ImageService {
         oci_path
     }
 
+    #[tokio::main(flavor = "current_thread")]
+    async fn download_image(
+        image: &str,
+        auth: &RegistryAuth,
+    ) -> anyhow::Result<(OciManifest, String, ImageData)> {
+        let reference = Reference::try_from(image)?;
+        let mut client = Client::default();
+        let (image_manifest, _image_digest, image_config) =
+            client.pull_manifest_and_config(&reference, auth).await?;
+
+        let mut last_error = None;
+        let mut image_data = ImageData {
+            layers: Vec::with_capacity(0),
+            digest: None,
+        };
+
+        for i in 1..2 {
+            match client
+                .pull(
+                    &reference,
+                    auth,
+                    vec![
+                        manifest::IMAGE_DOCKER_LAYER_GZIP_MEDIA_TYPE,
+                        manifest::IMAGE_LAYER_GZIP_MEDIA_TYPE,
+                        MEDIA_TYPE_LAYEE_GZIP_ENC,
+                        MEDIA_TYPE_LAYER_ENC,
+                        MEDIA_TYPE_LAYER_NON_DISTRIBUTABLE_ENC,
+                        MEDIA_TYPE_LAYER_NON_DISTRIBUTABLE_GZIP_ENC,
+                    ],
+                )
+                .await
+            {
+                Ok(data) => {
+                    image_data = data;
+                    break;
+                }
+                Err(e) => {
+                    println!(
+                        "Got error on pull call attempt {}. Will retry in 1s: {:?}",
+                        i, e
+                    );
+                    last_error.replace(e);
+                    tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+                }
+            }
+        }
+
+        Ok((image_manifest, image_config, image_data))
+    }
+
+    fn pull_image_with_oci_distribution(
+        image: &str,
+        cid: &str,
+        source_creds: &Option<String>,
+    ) -> Result<()> {
+        let oci_path = Self::build_oci_path(cid);
+        fs::create_dir_all(&oci_path)?;
+
+        info!(
+            sl!(),
+            "Attempting to pull image with rust crate {}...", image
+        );
+
+        let mut auth = RegistryAuth::Anonymous;
+        if let Some(source_creds) = source_creds {
+            let auth_info: Vec<&str> = source_creds.split(':').collect();
+            if auth_info.len() == 2 {
+                auth = RegistryAuth::Basic(auth_info[0].to_string(), auth_info[1].to_string());
+            } else {
+                warn!(sl!(), "invalid source_creds format \n");
+            }
+        }
+
+        let (mut image_manifest, image_config, image_data) = Self::download_image(image, &auth)?;
+
+        // Prepare OCI layout storage for umoci
+        image_manifest.config.media_type = manifest::IMAGE_CONFIG_MEDIA_TYPE.to_string();
+        let oci_blob_path = format!("{}/blobs/sha256/", oci_path.to_string_lossy());
```

Comment:
currently, we only support sha256. For to_string_lossy(), since oci_blob_patch is formated from oci_path which is a PathBuf struct. we may need keep oci_blob_patch as str since it will be used to format the filename to store file.

Reply:
Then I think we want `oci_blob_path` to be a `PathBuf` as well, built on top of `oci_path`.